prem_utils/connectors/prem.py

```diff
@@ -44,98 +41,78 @@ def parse_chunk(self, chunk):
-            response = client.post(url, json=request_data, headers=request_headers, timeout=600)
-            return response.json()
-
-    def _stream_generator_wrapper(self, url, request_data):
-        """
-        Wraps the async streaming generator to be consumed like a synchronous generator.
-        """
-        # Queue to hold streamed chunks
-        chunk_queue = queue.Queue()
-
-        def run_async():
-            asyncio.run(self._consume_streaming_endpoint(url, request_data, chunk_queue))
-
-        # Start the asynchronous generator in a separate thread
-        threading.Thread(target=run_async).start()
-
-        # Yield from queue in the current (synchronous) thread
-        while True:
-            chunk = chunk_queue.get()
-            if chunk is None:  # None is used as a signal to indicate completion
-                break
-            yield chunk
+        try:
+            if stream:
+                return self._stream_response(prem_ml_url, request_data)
+            else:
+                response = await self._perform_request(prem_ml_url, request_data)
+                return self._format_response(messages, response)
+        except Exception as e:
+            raise e
+
+    async def _perform_request(self, url, request_data):
+        headers = {"Authorization": f"Bearer {self.prem_ml_key}", "Content-Type": "application/json"}
+        async with httpx.AsyncClient() as client:
+            response = await client.post(url, json=request_data, headers=headers, timeout=600)
+        return response.json()
 
-    async def _consume_streaming_endpoint(self, url, request_data, chunk_queue):
-        """
-        Asynchronous generator to consume a streaming endpoint and put the content into a queue.
-        """
-        request_headers = {"Authorization": f"Bearer {self._api_key}", "Content-Type": "application/json"}
+    async def _stream_response(self, url, request_data):
+        headers = {"Authorization": f"Bearer {self.prem_ml_key}", "Content-Type": "application/json"}
         async with httpx.AsyncClient() as client:
-            async with client.stream("POST", url, json=request_data, headers=request_headers, timeout=600) as response:
+            async with client.stream("POST", url, json=request_data, headers=headers, timeout=600) as response:
+                buffer = ""
                 async for chunk in response.aiter_text():
-                    if len(chunk.strip()) != 0:
-                        # Sometimes (randomly) chunks contains multiple lines
-                        lines = chunk.strip().split("\n")
-                        for line in lines:
-                            chunk_queue.put(json.loads(line))
-        chunk_queue.put(None)  # Signal completion
+                    buffer += chunk
+                    while "\n" in buffer:
+                        line, buffer = buffer.split("\n", 1)
+                        if line.strip():
+                            yield json.loads(line)
```

Comment:
why do we need a buffer?

Reply:
Yea, not very intuitive, but in my experiments, I observed that sometimes the response from prem-ml yielded incomplete chunks. Due to this, the json parsing of the line would error out. To deal with this, I added a buffer here to cache in incomplete responses and put them together. 

This was an edge case that was observed only a few times during experimenting. 

I also checked the streaming function in prem-ml  ([link](https://github.com/premAI-io/prem-ml/blob/406105d5668ff04dcb6ff67625328f7ad97f3109/prem_ml/core/services.py#L150C4-L161C86)) and it yields only a token at a time. Couldn't understand why this was happening, so I decided to add a buffer in.

-------

`prem-ml stream_generator()`
```
    def stream_generator():
        try:
            for response in modal_model.completion.remote_gen(data, stream=True):
                yield json.dumps(prepare_chat_response_chunk(response=response["content"], model=model)) + "\n"
        except httpx.RequestError as error:
            yield json.dumps(
                {"status": error.response.status_code if error.response else None, "error": str(error)}
            ) + "\n"
        except ValueError as error:
            yield json.dumps({"status": 500, "error": str(error)}) + "\n"

    return StreamingHttpResponse(stream_generator(), content_type="application/json")
```